crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2016,12 +2016,50 @@ impl<'db> Type<'db> {
             | Type::FunctionLiteral(..) => {
                 let fallback = self.instance_member(db, name_str);
 
-                self.invoke_descriptor_protocol(
+                let result = self.invoke_descriptor_protocol(
                     db,
                     name_str,
                     fallback,
                     InstanceFallbackShadowsNonDataDescriptor::No,
-                )
+                );
+
+                let custom_getattr_result =
+                    || {
+                        // Typeshed has a fake `__getattr__` on `types.ModuleType` to help out with dynamic imports.
+                        // We explicitly hide it here to prevent arbitrary attributes from being available on modules.
+                        if self.into_instance().is_some_and(|instance| {
+                            instance.class.is_known(db, KnownClass::ModuleType)
+                        }) {
+                            return Symbol::Unbound.into();
+                        }
+
+                        self.try_call_dunder(
+                            db,
+                            "__getattr__",
+                            &CallArguments::positional([Type::StringLiteral(
+                                StringLiteralType::new(db, Box::from(name.as_str())),
+                            )]),
+                        )
+                        .map(|outcome| Symbol::bound(outcome.return_type(db)))
+                        // TODO: Handle call errors here.
+                        .unwrap_or(Symbol::Unbound)
```

Comment:
An interesting (and useful) behavior that I think falls out of this implementation, but isn't tested, is that if you annotate your `__getattr__` to accept e.g. a union of literal string types for the attribute name, then only those attribute names will fall back.

A less-useful corollary of this implementation is that if you make a dumb mistake in the signature of your `__getattr__` method, the only feedback you will get is attributes failing to resolve. Not sure if or how this could be helped. We don't need to worry about it for this PR, I think.